packages/openapi/src/middleware.ts

```diff
@@ -2,39 +2,74 @@ import { OpenAPI, RequestOptions, isValidationError } from './'
 
 import Koa from 'koa'
 
+interface ValidationOptions {
+  validateRequest?: boolean
+  validateResponse?: boolean
+}
+
 export function createValidatorMiddleware<T extends Koa.DefaultContext>(
   spec: OpenAPI,
-  options: RequestOptions
+  options: RequestOptions,
+  validationOptions: ValidationOptions | undefined = {
+    validateRequest: true,
+    validateResponse: true
+  }
 ): (ctx: Koa.Context, next: () => Promise<unknown>) => Promise<void> {
-  const validateRequest = spec.createRequestValidator<T['request']>(options)
-  const validateResponse =
-    process.env.NODE_ENV !== 'production' &&
-    spec.createResponseValidator(options)
+  const requestValidator = spec.createRequestValidator<T['request']>(options)
+  const responseValidator = spec.createResponseValidator(options)
 
   return async (
     ctx: Koa.Context,
     next: () => Promise<unknown>
   ): Promise<void> => {
-    // TODO: Allow 'application/*+json'
-    ctx.assert(ctx.accepts('application/json'), 406, 'must accept json')
     try {
-      if (validateRequest(ctx.request)) {
-        await next()
-        if (validateResponse && !validateResponse(ctx.response)) {
-          throw new Error('unreachable')
+      if (validationOptions?.validateRequest) {
+        // TODO: Allow 'application/*+json'
+        if (!ctx.accepts('application/json')) {
+          throw new OpenAPIValidatorMiddlewareError(
+            'Received error validating OpenAPI request: Must accept application/json',
+            406
+          )
         }
-      } else {
-        throw new Error('unreachable')
+
+        requestValidator(ctx.request)
       }
     } catch (err) {
-      if (err instanceof Koa.HttpError) {
-        throw err
-      } else if (isValidationError(err)) {
-        ctx.throw(err.status ?? 500, err.errors[0])
-      } else {
-        console.log('err=', err)
-        ctx.throw(500)
+      if (isValidationError(err)) {
+        throw new OpenAPIValidatorMiddlewareError(
+          `Received error validating OpenAPI request: ${err.errors[0]?.message}`,
+          err.status || 500
+        )
       }
+
+      throw err // Should not be possible (only ValidationError is thrown in requestValidator)
     }
```

Comment:
I think this is functionally equivalent but perhaps a bit more explicit? only try/catching `requestValidator` basically.

```ts
    if (validationOptions?.validateRequest) {
      if (!ctx.accepts('application/json')) {
        throw new OpenAPIValidatorMiddlewareError(
          'Received error validating OpenAPI request: Must accept application/json',
          406
        )
      }
      try {
        requestValidator(ctx.request)
      } catch (err) {
        if (isValidationError(err)) {
          throw new OpenAPIValidatorMiddlewareError(
            `Received error validating OpenAPI request: ${err.errors[0]?.message}`,
            err.status || 500
          )
        }

        throw err // Should not be possible (only ValidationError is thrown in requestValidator)
      }
    }
```

As-is, the 406 `OpenAPIValidatorMiddlewareError` is being re-thrown where the comment suggests it should be impossible. I guess in that case we could just remove the comment, but it seems like catching it is unnecessary in the first place.

Reply:
Yup, definitely good call. Updating.